### PR TITLE
[RUN-5033] Windowing API calls nack if size constraints would be violated

### PR DESCRIPTION
**Pull request by MichaelMCoates**
#### Description of Change
This PR modifies Windowing APIs that resize windows. If a Windowing API call attempts to resize a window outside of its `maxHeight`/`maxWidth`/`minHeight`/`minWidth` constraints, the function will nack. 

Affected functions:
`Window.animate`
`Window.resizeBy`
`Window.resizeTo`
`Window.setBounds`

**This PR will change the behavior of these API calls. Previously, the windows would resize up to their constraints, and the functions would ack. Now, we prevent all resizing, and nack instead**

The following tests will need to have a Max Version set @whyn07m3 :
https://testing-dashboard.openfin.co/#/app/tests/55897c432517bd926e4a257b/edit
https://testing-dashboard.openfin.co/#/app/tests/55897c432517bd926e4a257c/edit

The following tests will need to have a Min Version set (once this PR is merged and we've got a build number):
https://testing-dashboard.openfin.co/#/app/tests/5cf15765cb360141a7dfe572/edit
https://testing-dashboard.openfin.co/#/app/tests/5cf1515ecb360141a7dfe571/edit
https://testing-dashboard.openfin.co/#/app/tests/5ce46a6bcb360141a7dfe43a/edit
https://testing-dashboard.openfin.co/#/app/tests/5ce44742cb360141a7dfe42d/edit
https://testing-dashboard.openfin.co/#/app/tests/5ce441cecb360141a7dfe427/edit
https://testing-dashboard.openfin.co/#/app/tests/5ce44129cb360141a7dfe426/edit
https://testing-dashboard.openfin.co/#/app/tests/5ce31e0fcb360141a7dfe411/edit

#### Checklist

- [X] PR description included and stakeholders cc'd
- [X] `npm test` passes
- [X] automated tests are [changed or added](https://testing-dashboard.openfin.co/#/app/dashboard)
- [?] manual tests are [changed or added](https://github.com/openfin/test_apps)
- [?] relevant documentation is [changed or added](https://github.com/hadoukenio/js-adapter)
- [X] PR title starts with the JIRA ticket [pull request process](https://github.com/openfin/Internal-Wiki/wiki/Pull-Request-Process)
- [ ] PR release notes describe the change in a way relevant to app-developers

#### Release Notes
Windowing APIs that affect window size will now nack if that resize action would violate the window's size constraints.

Notes:

### CODE REVIEW

**Comment by MichaelMCoates** [comment]
PR Ready for review. I could use help with cleanup; I've got a lot of code duplication but not sure the best way to handle it.

File: src/browser/api/window.js [hunk 1]
```diff
@@ -1074,7 +1074,35 @@ Window.animate = function(identity, transitions, options = {}, callback = () =>
         animationMeta.interrupt = true;
     }
 
-    animations.getAnimationHandler().add(browserWindow, animationMeta, animationTween, callback, errorCallback);
+    const { size } = transitions;
+
+    if (!size) {
+        animations.getAnimationHandler().add(browserWindow, animationMeta, animationTween, callback, errorCallback);
```
Comment (tomer-openfin) [inline src/browser/api/window.js]: what's Tween?

File: src/browser/api/window.js [hunk 2]
```diff
@@ -1558,24 +1586,81 @@ Window.removeEventListener = function(identity, type, listener) {
...
+        aspectRatio
+    } = options;
+
+    if (width === undefined && height === undefined) {
+        return true;
+    }
 
-Window.resizeBy = function(identity, deltaWidth, deltaHeight, anchor) {
+    if (height === undefined) {
+        return (width >= minWidth) && (maxWidth === -1 || width <= maxWidth);
```
Comment (tomer-openfin) [inline src/browser/api/window.js]: how about maxWidth\Height <=0 here? I'm guessing users who want to remove max/min limitation would set it's value to 0.

File: src/browser/api/window.js [hunk 3]
```diff
@@ -1558,24 +1586,81 @@ Window.removeEventListener = function(identity, type, listener) {
...
+    }
+
+    if (width === undefined) {
+        return (height >= minHeight) && (maxHeight === -1 || height <= maxHeight);
+    }
+
+    const acceptableWidth = (width >= minWidth) && (maxWidth === -1 || width <= maxWidth);
+    const acceptableHeight = (height >= minHeight) && (maxHeight === -1 || height <= maxHeight);
+    const roundedProposedRatio = Math.round(100 * (width / height)) / 100;
```
Comment (tomer-openfin) [inline src/browser/api/window.js]: what's the motivation for rounding here? get a 2 decimal point precision? 
I'm not a big fan of comments, but I think this bit is technical enough to justify including some help for future readers

File: src/browser/api/window.js [hunk 4]
```diff
@@ -1558,24 +1586,81 @@ Window.removeEventListener = function(identity, type, listener) {
...
+
+Window.resizeBy = function(identity, deltaWidth, deltaHeight, anchor, callback, errorCallback) {
     const browserWindow = getElectronBrowserWindow(identity);
     const opts = { anchor, deltaHeight, deltaWidth };
     if (!browserWindow) {
         return;
     }
-    NativeWindow.resizeBy(browserWindow, opts);
+
+    if (!('_options' in browserWindow)) {
```
Comment (tomer-openfin) [inline src/browser/api/window.js]: Does anyone know if the middleware is testing for this?

**Comment by tomer-openfin** [comment]
Question - in the following scenario: 
`Two windows are grouped and snapped. One of them has a minHeight. Someone is setting bounds on one of these windows to a lower value than the minHeight.`

For both the current implementation and ideally -

- would the call just nack and no change will occur *or* would the call ack with just 1 window changing? or a 3rd option?

- does it matter if the getBounds call is executed from the window with the minHeight or not?

**Comment by whyn07m3** [comment]
@MichaelMCoates please, remind me to set max version for automated tests once this is merged and has a build number

**Comment by whyn07m3** [comment]
@MichaelMCoates
image
the link above is not pointing to added or changed tests, please list the tests that were added or changed

File: src/browser/api/window.js (hunk 4 above)
Comment (whyn07m3) [reply src/browser/api/window.js]: Btw, this is being done a few times, good candidate for DRYing

File: src/browser/api/window.js [hunk 5]
```diff
@@ -1558,24 +1586,81 @@ Window.removeEventListener = function(identity, type, listener) {
...
+        NativeWindow.resizeBy(browserWindow, opts);
+        callback();
+    } else {
+        errorCallback(new Error(`Proposed window bounds violate size constraints for uuid: ${identity.uuid} name: ${identity.name}`));
+    }
 };
 
 
-Window.resizeTo = function(identity, width, height, anchor) {
+Window.resizeTo = function(identity, width, height, anchor, callback, errorCallback) {
```
Comment (whyn07m3) [inline src/browser/api/window.js]: Since these functions are synchronous, and I think we take care of this case, where we can simply throw an error instead of calling an error callback... not sure what's best.

**Comment by MichaelMCoates** [comment]
> @MichaelMCoates
> image
> the link above is not pointing to added or changed tests, please list the tests that were added or changed. Same for other items in the list

Thanks, didn't realize that private tests were visible to others! Added.

File: src/browser/api/window.js (hunk 1 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: https://cdn.openfin.co/jsdocs/stable/tutorial-window.animate.html

Different movement styles, essentially.

File: src/browser/api/window.js (hunk 3 above)
Comment (dhamberlin) [reply src/browser/api/window.js]: Shoutout to my man `Number.toFixed`. 

Some pros/cons vs this method: https://stackoverflow.com/a/12830454
tl;dr: returns a string, so you probably need to change it back to a number, but it handles rounding better in some cases.

^ this is probably a neckbeard nitpick and either version is probably fine for our purposes.

File: src/browser/api/window.js (hunk 2 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: Anybody got info on what happens when you set maxWidth/Height to 0 vs -1? Seems like most of the code is expecting -1.

**Comment by pbaize** [comment]
How about making a validation decorator function that just decorates the methods rather than changing each of them?

File: src/browser/api/window.js (hunk 2 above)
Comment (pbaize) [reply src/browser/api/window.js]: https://appoji.jira.com/browse/RUN-5019

File: src/browser/api/window.js (hunk 3 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: Yeah, 2 decimal point precision. Ran a bunch of tests to see how the `aspectRatio` calculation changes depending on width/height, and it seems that our current manual resizing logic has 2-decimal precision. Is this too precise? Not precise enough? Adding comment.

File: src/browser/api/window.js [hunk 6]
```diff
@@ -1558,24 +1586,81 @@ Window.removeEventListener = function(identity, type, listener) {
...
+function areNewBoundsWithinConstraints(options, width, height) {
+    const {
+        minWidth,
+        minHeight,
+        maxWidth,
+        maxHeight,
+        aspectRatio
+    } = options;
+
+    if (width === undefined && height === undefined) {
```
[the remark this answers is not in the document]
Comment (MichaelMCoates) [reply src/browser/api/window.js]: Good call, changing.

File: src/browser/api/window.js (hunk 2 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: With that in mind, should I leave as is then?

File: src/browser/api/window.js (hunk 3 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: @dhamberlin oooooooo really cool, but looking at some of the comments underneath, seems like `Number.toFixed` also has rounding weirdness sometimes. This one in particular caught my eye, what do you think?:

> There is no advantage of using toFixed instead of Math.round; toFixed leads to quite the same rounding problems (try them with 5.555 and 1.005), but is like 500x (no kidding) slower than Math.round – Pierre May 20 '14 at 16:49

Comment (dhamberlin) [reply src/browser/api/window.js]: Good looking out. I like `toFixed` in that its intent is clear (re: Tomer's comment). But this is one of the places in our code base where speed at this scale is liable to actually matter. Haven't benchmarked, but if `Math.round` is faster, we should probably use that.

Comment (dhamberlin) [reply src/browser/api/window.js]: Follow up cause talking perf without testing squicks me out:

Ran some tests and this Pierre fellow is right. My results were a little less dramatic than 500x, but `toFixed` was decidedly slower (took about 2-3 times as long as `Math.round(100 * n) / 100` for most of my test runs)

File: src/browser/api/window.js (hunk 2 above)
Comment (tomer-openfin) [reply src/browser/api/window.js]: as I see it - some users will try passing falsy values in an attempt to cancel a min/max, let's save them the trip to the docs.

File: src/browser/api/window.js (hunk 5 above)
Comment (MichaelMCoates) [reply src/browser/api/window.js]: I was following the convention of the rest of the file, maybe we follow the convention in this PR and move it to error throwing in a separate PR?

**Comment by MichaelMCoates** [comment]
> Question - in the following scenario:
> `Two windows are grouped and snapped. One of them has a minHeight. Someone is setting bounds on one of these windows to a lower value than the minHeight.`
> 
> For both the current implementation and ideally -
> 
> * would the call just nack and no change will occur _or_ would the call ack with just 1 window changing? or a 3rd option?
> * does it matter if the getBounds call is executed from the window with the minHeight or not?

Just tested on my machine. Created two windows, snapped them together, set a `maxHeight` of 600 pixels on one window (Window A), and tried to resize past 600px from both windows programmatically.

**9.61.38.43**
-> Resizing Window A: Window A resizes up to its constraint. Window B resizes up to the height set in `resizeTo`. Window B's width resizes strangely.
-> Resizing Window B: Window A resizes up to its constraint. Window B resizes to the bounds set in `resizeTo`

**11.69.42.19**
It seems that all resizing APIs don't do anything. The windows don't move, and there are no errors. 

**This PR**
Same as `11.69.42.19`. Because we hijack all move/resize API calls from grouped windows, we don't hit any of my code.

How should we handle this?

**Comment by MichaelMCoates** [comment]
Looping in @datamadic @tgoc99 . Can you take a look at the above please?

**Comment by tgoc99** [comment]
Looks good... From a grouped move perspective, it will need to be handled elsewhere.

**Comment by tomer-openfin** [comment]
Thanks for running these tests! 

> **11.69.42.19**
> It seems that all resizing APIs don't do anything. The windows don't move, and there are no errors (which is bad, right?).
> 
> **This PR**
> Same as `11.69.42.19`. Because we hijack all move/resize API calls from grouped windows, we don't hit any of my code.

Does this probably deserve another ticket @datamadic?

**Comment by tgoc99** [comment]
> Thanks for running these tests!
> 
> > **11.69.42.19**
> > It seems that all resizing APIs don't do anything. The windows don't move, and there are no errors (which is bad, right?).
> > **This PR**
> > Same as `11.69.42.19`. Because we hijack all move/resize API calls from grouped windows, we don't hit any of my code.
> 
> Does this deserve another ticket @datamadic?

Will be fixed with the new native / external window grouping PR.
